### ISSUE TITLE
Setting values on Y.Map while in a deleted state results in values being undefined when deletion is undone

### ISSUE THREAD

**Issue by pierscowburn** [yjs/yjs#472]
**Checklist**

* [x] Are you reporting a bug? Use github issues for bug reports and feature requests. For general questions, please use https://discuss.yjs.dev/
* [x] Try to report your issue in the correct repository. Yjs consists of many modules. When in doubt, report it to https://github.com/yjs/yjs/issues/

**Describe the bug**
I'm running into an issues that feels similar to https://github.com/yjs/yjs/issues/443, but is not addressed by the [fix](https://github.com/yjs/yjs/commit/df2d59e2fb28c10e57a5615c772a29ca27cf98e1) in 13.5.41.

What I observe is that setting a value on a Y.Map when the map is currently in a deleted state (i.e. because it was literally deleted, or because its creation was undone) results in the corresponding value being undefined in the case where the map is subsequently un-deleted (i.e. via undo/redo). The behaviour also seems to be non-deterministic, as far as I've been able to tell from writing a minimal example.

**To Reproduce**
_First example: Undoing the creation of a map, setting a field, then redoing the creation_ 
1. Create a test setup emulating two users connected to each other via websocket.
2. Have user A create a Y.Map, and create a field on it
3. Have user B retrieve a reference to the map
4. Have user A undo the creation of the map 
5. Have user B change the value of the field, using the reference created in (3)
6. Have user A redo the creation of the map
7. Attempt to retrieve the field via either user – the value will be undefined on roughly 50% of runs

_Second example: Deleting a map, setting a field, then undoing the deletion_ 
1. Create a test setup emulating two users connected to each other via websocket.
2. Have user A create a Y.Map, and create a field on it
3. Have user B retrieve a reference to the map
4. Have user A delete the map
5. Have user B change the value of the field, using the reference created in (3)
6. Have user A undo the deletion of the map
7. Attempt to retrieve the field via either user – the value will be undefined on roughly 50% of runs

**Expected behavior**
The field should exist and either have the value set in (1) or the value set in (5), rather than being undefined. It seems that it's the setting of the field that causes the deletion of it – fields that are not touched while the map is in the deleted state are still intact after the deletion is undone.

**Example Code**
The following code demonstrates the issue:

```typescript
import * as Y from 'yjs';
import { WebsocketProvider } from 'y-websocket';
import WebSocket from 'ws';
import chalk from 'chalk';
import fs from 'fs';

// @ts-expect-error Need to provide WebSocket for y-websocket, as we're not running in a browser
global.WebSocket = WebSocket;

type Results = {
  example1: {
    userAFinalValue?: string,
    userBFinalValue?: string,
  },
  example2: {
    userAFinalValue?: string,
    userBFinalValue?: string,
  },
};

const runTest = () => {
  const seed = Date.now() + ':' + Math.random(); // Used for generating a unique room name for this test pass

  const results: Results = {
    example1: {},
    example2: {},
  };

  const createContext = (name: string) => {
    const ydoc = new Y.Doc();
  
    const root = ydoc.getMap('root');
  
    const websocketProvider = new WebsocketProvider('ws://localhost:1234', name + seed, ydoc);
  
    const undoManager = new Y.UndoManager(root, { captureTimeout: 0 });

    return {
      ydoc,
      root,
      websocketProvider,
      undoManager,
      destroy() {
        websocketProvider.disconnect();
        websocketProvider.destroy();
        undoManager.destroy();
        ydoc.destroy();
      },
    };
  };
  
  {
    const userAContext = createContext('example1');
    const userBContext = createContext('example1');
    
    let userANestedMap: Y.Map<unknown>;
    let userBNestedMap: Y.Map<unknown>;
    
  
    console.log('YJS Version:', JSON.parse(fs.readFileSync('node_modules/yjs/package.json').toString()).version);
    
    // ------------------------------------------------------------------------
    
    console.log(chalk.bold.yellow('\nFirst example: Undoing the creation of a map, setting a field, then redoing the creation'));
    
    // ------------------------------------------------------------------------
    
    
    // ------------------------------------------------------------------------
    console.log(chalk.bold.cyan('\nInitial setup'));
    // ------------------------------------------------------------------------
    
    userANestedMap = new Y.Map();
    userAContext.root.set('nested', userANestedMap);
    
    userBNestedMap = userBContext.root.get('nested') as Y.Map<unknown>;
    userBNestedMap.set('foo', 10);
    userBNestedMap.set('bar', 20); // Will be used at the end for comparison
    
    console.log(' [userA] nested.foo', userANestedMap.get('foo')); // 10
    console.log(' [userB] nested.foo', userBNestedMap.get('foo')); // 10
    
    
    // ------------------------------------------------------------------------
    console.log(chalk.bold.cyan('\nHappy case test'));
    // ------------------------------------------------------------------------
    
    userAContext.undoManager.undo(); // Undo the creation of the nested map
    
    console.log(' After undoing map creation (both should be undefined):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // undefined (as expected)
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // undefined (as expected)
    
    userAContext.undoManager.redo(); // Redo the creation of the nested map
    
    userANestedMap = userAContext.root.get('nested') as Y.Map<unknown>;
    
    userBNestedMap = userBContext.root.get('nested') as Y.Map<unknown>;
      
    console.log(' After redoing map creation (both should be 10):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // 10
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // 10
    
    
    // ------------------------------------------------------------------------
    console.log(chalk.bold.cyan('\nUnhappy case test'));
    // ------------------------------------------------------------------------
    
    userAContext.undoManager.undo(); // Undo the creation of the nested map again
    
    console.log(' After undoing map creation (both should be undefined):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // undefined (as expected)
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // undefined (as expected)
    
    userBNestedMap.set('foo', 11); // Set foo, while the map creation is still undone
    
    userAContext.undoManager.redo(); // Redo the creation of the nested map
    
    userANestedMap = userAContext.root.get('nested') as Y.Map<unknown>;
    userBNestedMap = userBContext.root.get('nested') as Y.Map<unknown>;
      
    console.log(' After redoing map creation (both should be either 10 or 11, but not undefined):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // undefined (should either be 10 or 11, but not undefined)
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // undefined (should either be 10 or 11, but not undefined)

    results.example1.userAFinalValue = userANestedMap.get('foo') as string | undefined;
    results.example1.userBFinalValue = userBNestedMap.get('foo') as string | undefined;
      
    console.log(' For comparison, the values of \'bar\' are still intact:');
    console.log('  [userA] nested.bar', userANestedMap.get('bar')); // 20 - bar is intact, because no attempt to set it was made while the map creation was in the undone state
    console.log('  [userB] nested.bar', userBNestedMap.get('bar')); // 20 - bar is intact, because no attempt to set it was made while the map creation was in the undone state
  
    userAContext.destroy();
    userBContext.destroy();
  }
  
  
  
  {
    const userAContext = createContext('example2');
    const userBContext = createContext('example2');
    
    let userANestedMap: Y.Map<unknown>;
    let userBNestedMap: Y.Map<unknown>;
  
  
    // ------------------------------------------------------------------------
  
    console.log(chalk.bold.yellow('\n\nSecond example: Deleting a map, setting a field, then undoing the deletion'));
  
    // ------------------------------------------------------------------------
  
  
    // ------------------------------------------------------------------------
    console.log(chalk.bold.cyan('\nInitial setup'));
    // ------------------------------------------------------------------------
  
    userANestedMap = new Y.Map();
    userAContext.root.set('nested', userANestedMap);
  
    userBNestedMap = userBContext.root.get('nested') as Y.Map<unknown>;
    userBNestedMap.set('foo', 10);
    userBNestedMap.set('bar', 20); // Will be used at the end for comparison
  
    console.log(' [userA] nested.foo', userANestedMap.get('foo')); // 10
    console.log(' [userB] nested.foo', userBNestedMap.get('foo')); // 10
  
  
    // ------------------------------------------------------------------------
    console.log(chalk.bold.cyan('\nHappy case test'));
    // ------------------------------------------------------------------------
  
    userAContext.root.delete('nested'); // Delete the nested map
  
    console.log(' After deleting the map (both should be undefined):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // undefined (as expected)
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // undefined (as expected)
  
    userAContext.undoManager.undo(); // Undo the deletion of the nested map
  
    userANestedMap = userAContext.root.get('nested') as Y.Map<unknown>;
  
    userBNestedMap = userBContext.root.get('nested') as Y.Map<unknown>;
      
    console.log(' After undoing map deletion (both should be 10):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // 10
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // 10
  
  
    // ------------------------------------------------------------------------
    console.log(chalk.bold.cyan('\nUnhappy case test'));
    // ------------------------------------------------------------------------
  
    userAContext.root.delete('nested'); // Delete the nested map again
  
    console.log(' After deleting the map (both should be undefined):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // undefined (as expected)
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // undefined (as expected)
  
    userBNestedMap.set('foo', 11); // Set foo, while the map is still in the deleted state
  
    userAContext.undoManager.undo(); // Undo the deletion of the nested map
  
    userANestedMap = userAContext.root.get('nested') as Y.Map<unknown>;
    userBNestedMap = userBContext.root.get('nested') as Y.Map<unknown>;
      
    console.log(' After undoing map deletion (both should be either 10 or 11, but not undefined):');
    console.log('  [userA] nested.foo', userANestedMap.get('foo')); // undefined (should either be 10 or 11, but not undefined)
    console.log('  [userB] nested.foo', userBNestedMap.get('foo')); // undefined (should either be 10 or 11, but not undefined)
    
    results.example2.userAFinalValue = userANestedMap.get('foo') as string | undefined;
    results.example2.userBFinalValue = userBNestedMap.get('foo') as string | undefined;
      
    console.log(' For comparison, the values of \'bar\' are still intact:');
    console.log('  [userA] nested.bar', userANestedMap.get('bar')); // 20 - bar is intact, because no attempt to set it was made while the map creation was in the undone state
    console.log('  [userB] nested.bar', userBNestedMap.get('bar')); // 20 - bar is intact, because no attempt to set it was made while the map creation was in the undone state
  
    userAContext.destroy();
    userBContext.destroy();
  }

  return results;
};

const allResults: Results[] = [];
const numRuns = 100;

for (let i = 0; i < numRuns; ++i) {
  allResults.push(runTest());
}

console.log(allResults);

// ------------------------------------------------------------------------

console.log(chalk.bold.yellow('\n\nAggregate results:\n'));

// ------------------------------------------------------------------------

let numFailuresForExample1 = 0;
let numFailuresForExample2 = 0;

allResults.forEach(results => {
  if (results.example1.userAFinalValue === undefined || results.example1.userBFinalValue === undefined) {
    numFailuresForExample1++;
  }
  if (results.example2.userAFinalValue === undefined || results.example2.userBFinalValue === undefined) {
    numFailuresForExample2++;
  }
});

console.log(' Example 1 failures: ' + numFailuresForExample1 + ' / ' + numRuns);
console.log(' Example 2 failures: ' + numFailuresForExample2 + ' / ' + numRuns);
```

```json
{
  "name": "undo-redo-bug-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "start": "vite-node index.ts"
  },
  "dependencies": {
    "@types/ws": "8.5.3",
    "chalk": "5.1.0",
    "vite-node": "0.23.4",
    "ws": "8.9.0",
    "y-websocket": "1.4.5",
    "yjs": "13.5.41"
  }
}

```

To run:

```bash
npm start
```

Output:

The script will run the tests 100 times, printing verbose output for each run, plus a final set of aggregated results. The precise numbers change each time, but an example from my machine just now is as follows:

```
Aggregate results:
 Example 1 failures: 41 / 100
 Example 2 failures: 54 / 100
```

**Environment Information**
 - Node v16.14.2, also Chrome v105.0.5195.125
 - `"yjs": "13.5.41"`
 - `"y-websocket": "1.4.5"`

**Comment by dmonad** [comment]
Hi @pierscowburn,

I haven't read the full code, but based on the explanation, it sounds like everything is working as expected. 

In Yjs (the same is true for most CRDTs), you can't really undo a deletion. Items (like shared Y.* types) can be marked as deleted, but you can't just toggle that boolean to "undeleted" without causing major issues. So when you undo the deletion of a shared type, the `Y.UndoManager` creates a new object that looks exactly like the old one. However, it is technically a new object with a new reference.

Once a type is deleted, you are not supposed to insert content. That should be a no-op. I'm trying to keep the codebase small. Hence I don't throw errors when an action is performed that doesn't lead to inconsistencies. However, I didn't expect people to insert content into deleted types to break the undo-manager. So maybe I should throw an error instead.

To enable this "feature" to insert content into deleted types so that you can redo it later, I'd need to cripple the garbage-collector and make sure that we don't throw away information that is needed to restore the inserted content. Because I don't understand why this is necessary (it's probably a very unique use case), I'm currently favoring closing this ticket. Sorry for that... 

Did I misunderstand something?

**Comment by pierscowburn** [comment]
Thanks for the thoughtful reply. Here's why this is an issue for our application:

- The scenario whereby one user can set a property on an object that has been deleted (via undo or otherwise) can happen quite easily in cases where one or both of the users has a poor/intermittent connection, or is working offline for a while.
- At the point at which the object is then re-created, because of the behaviour I described above, the object is then invalid - there are certain properties of objects that our schema expects to exist, and our code ensures that they always do exist via all of the workflows that are under our control. The codepath of object creation undone -> property of object set by another user -> object creation redone is a path via which objects can be created, but which we don't directly control – and the resulting objects are being created in an invalid state due to this behaviour.

> That should be a no-op.

I agree that it should be a no-op – I wouldn't describe what's happening here as a no-op though. A no-op would be the property still being there in its original state when the object is recreated (i.e. the value should be `10` in the above code). Instead the property is `undefined`, whereas sibling properties that were not set while the object was deleted are still intact.

> To enable this "feature" to insert content into deleted types so that you can redo it later,

Apologies for any confusion: this isn't what's being proposed. All I'm saying is that the data should be unaffected by the attempts to set fields while the objects are deleted (i.e. a no-op, as you say).

**Comment by dmonad** [comment]
```javascript
import * as Y from 'yjs'

for (let i = 0; i < 100; i++) {
  const ydoc1 = new Y.Doc()
  const ydoc2 = new Y.Doc()

  const sync = () => {
    Y.applyUpdate(ydoc1, Y.encodeStateAsUpdate(ydoc2))
    Y.applyUpdate(ydoc2, Y.encodeStateAsUpdate(ydoc1))
  }

  const root1 = ydoc1.getMap('root')
  const root2 = ydoc2.getMap('root')

  const um1 = new Y.UndoManager(root1)

  const ymap1 = new Y.Map()
  // 1. Have user A create a Y.Map, and create a field on it
  root1.set('ymap', ymap1)
  ymap1.set('key', 'value1')

  // 2. Have user B retrieve a reference to the map
  sync()
  const ymap2 = root2.get('ymap')

  // 3. Have user A undo the creation of the map
  um1.undo()

  // 4. Have user B change the value of the field, using the reference created in (3)
  ymap2.set('key', 'value2')

  // 5. Have user A redo the creation of the map
  um1.redo()
  sync()

  // 6. Attempt to retrieve the field via either user – the value will be undefined on roughly 50% of runs
  const result1 = root1.get('ymap').get('key')
  const result2 = root2.get('ymap').get('key')

  console.log({
    result1,
    result2
  })
}

```

When I do this, I always get the same result `"value1"`.

I suspect that your code the second user manipulates `ymap` when it is already deleted. However, that should be prevented.

**Comment by pierscowburn** [comment]
Thanks for the quick response!

I definitely think there's an element of latency that is factoring into the behaviour I'm seeing; that's the only way I can account for the non-determinism. My example above is using y-websocket to synchronise the changes rather than synchronizing explicitly, so it's possible that some of the sync steps are happening in a different order on the occasions where the value becomes undefined. In your example, if I insert a `sync()` call after step 4, the value does become undefined, so I imagine this is what's happening.

Would it not be a straightforward fix for the library to check whether the map is deleted at the point that the `set()` call happens, and if so, do nothing?

**Comment by pierscowburn** [comment]
I've been trying to find a way to work around this from our own code, but unfortunately I don't actually think it's possible to do so. Here are some notes on our setup and the way in which this issue is being triggered, for context:

- We're building a collaborative image editor/content creation app. As part of this we have the notion of layers (i.e. images, etc) that can be added to a composition and positioned by the user in a 2D viewport, via dragging with the mouse.
- The data for each layer is represented by a `SharedMap`.
- When a layer is dragged, a `transform` field on the map that represents the layer is updated to reflect the new 2D transform.
- Given this setup, one way in which the issue can be reliably triggered is:
  - User A imports an image
  - User B (on another machine) starts dragging the image in order to position it, resulting in a stream of changes to the `transform` field
  - User A undos their import action whilst User B's drag is in progress
  - User A redos their import action
  - The `transform` field on the layer's map is now undefined, because it was mutated while the layer was in its undone state

Things I've tried:
- Updating our code in order to prevent calls to set the `transform` (or any other field) if the layer has been deleted. This unfortunately does not resolve the issue, because some of the stream of continuous `transform` changes being sent from User B to User A are already in flight at the point of User A pressing undo – and so the mutations that take place when the layer is in its undone state are being applied internally by YJS on User A's machine in response to the in-flight changes being received, as opposed to being applied by our own code.
- Patching `SharedMap.set()` to check for `this._item?.deleted` and bail if this evaluates to true. This also does not resolve the issue, for the same reason.
- Patching `integrateStructs()` to avoid the [`stackHead.integrate()` and subsequent `state.set()` calls](https://github.com/yjs/yjs/blob/2ec19defcb244fa32d0033f050c38c0f57a01e59/src/utils/encoding.js#L331-L332) in the case where `stackHead.parent?._item?.deleted` is true. I'm aware that this is not a valid thing to do (and YJS's own tests fail when I apply this patch, for good reason) – but I thought that it might be useful context to know that having this patch in place does reliably resolve the actual bug that we observe in our own application.

I'd love to get some guidance on whether there is a valid patch that could be applied to the library to help us avoid this issue. Please let me know if I can provide any more info/context.

**Comment by astnmsn** [comment]
From the bit of digging I have done with this example, it appears that the behavior is a result of [this bit of code in Item.js](https://github.com/yjs/yjs/blob/ea7ad07f34cc64cbd5d30e3fe00fa2c5476dec9f/src/structs/Item.js#L214-L218):
```javascript
if (left && left.right !== null) {
    // It is not possible to redo this item because it conflicts with a
    // change from another client
    return null
}
```
The single call to `undoManager.redo()` will trigger a call to [redoItem()](https://github.com/yjs/yjs/blob/ea7ad07f34cc64cbd5d30e3fe00fa2c5476dec9f/src/structs/Item.js#L136) for 2 items. 
Using the naming from Kevin's example: 

1. An empty YMap item with the parentSub=`'ymap'` <- (this is the parent being restored)
2. `ContentAny` item where the content is `['value1']` and the parentSub=`'key'` <- (this is the child being restored with the original value set for `'key'`)

The first item is integrated into the transaction, but the second item is not.
The second call has a `left.right` with content = `ContentDeleted`.
This is the result of applying the update ('key'->'value2') from `root2` to `root1` after `root1` had already undone the creation of the parent. 

In previous versions of yjs, this check was done by comparing the item's `client.id` to the document's `clientID`, not by looking at `left.right` ([see here for the change](https://github.com/yjs/yjs/commit/4cfa49d601743e4366109b6e800125322b5c147a?diff=unified#diff-8e03df4677fec8fdd3103498dd8f3114030b2d18bca59ac692a122795103ab82L160)). This would have evaluated to false in this example and resulted in a subsequent call to `integrate()` the `['value1']` item.

It would seem that it is possible to add the `['value1']` item to the end of the original chain in much the same way that is done when other items to the right are to be [deleted](https://github.com/yjs/yjs/blob/ea7ad07f34cc64cbd5d30e3fe00fa2c5476dec9f/src/structs/Item.js#L204-L208):

```javascript
// Iterate right while right is in itemsToDelete
// If it is intended to delete right while item is redone, we can expect that item should replace right.
while (left !== null && left.right !== null && isDeleted(itemsToDelete, left.right.id)) {
    left = left.right
}
```

In this case the `ContentDeleted` item is not part of the `itemsToDelete` set. But following the comment in that code, it appears that `['value1']` could replace to the right on any `ContentDeleted` items that are referencing the same parentSub (because the intention is to delete those items as part of restoring `['value1']`). This would result in a chain that looks like `['value1']` -> `ContentDeleted` -> `['value1']` -> `null`, where the second `['value1']` is a new Item.

Whether or not this is completely safe to do is likely only answerable by Kevin. I don't immediately see anything that would prevent this, but I have only been looking at it for a short time, and on some small examples. It is likely safer when approaching it from this direction; as opposed to preventing the `set()` from doing it's normal operations when applying updates from `root2` to `root1`.

FWIW, If I were the user observing `root1`, I would expect the undo to restore the `key`'s value, as I was never even able to observe a change made by another user. 

Also, the `redoItem()` call internal to `Item` is called for both `undo()` and `redo()` operations in the UndoManager. So it makes sense that this would cause both behaviors that Piers is describing (redoing an undone create, or undoing a delete).

**Comment by astnmsn** [comment]
@pierscowburn I think for now your test should pass if you set `ignoreRemoteMapChanges` to true when instantiating the UndoManager. Although I cannot say if that will break other behavior you are relying on.

**Comment by pierscowburn** [comment]
@astnmsn thanks a lot for taking a look at this! I can confirm that enabling `ignoreRemoteMapChanges` does resolve the issue in our codebase, and also passes our existing test suite (which includes substantial tests for undo/redo behaviour).

I do still think that the original reported behaviour constitutes a bug, and will be unexpected by consumers of the library. As such I'm going to leave this open until @dmonad has chance to read the most recent few comments, in case there are any changes he wants to incorporate.

Thanks again!
